lib/ansible/cli/doc.py

```diff
@@ -63,7 +65,196 @@ class PluginNotFound(Exception):
     pass
 
 
-class DocCLI(CLI):
+class RoleMixin(object):
+    """A mixin containing all methods relevant to role argument specification functionality.
+
+    Note: The methods for actual display of role data are not present here.
+    """
+
+    ROLE_ARGSPEC_FILE = 'argument_specs.yml'
+
+    def _load_argspec(self, role_name, collection_path=None, role_path=None):
+        if collection_path:
+            path = os.path.join(collection_path, 'roles', role_name, 'meta', self.ROLE_ARGSPEC_FILE)
+        elif role_path:
+            path = os.path.join(role_path, 'meta', self.ROLE_ARGSPEC_FILE)
+        else:
+            raise AnsibleError("A path is required to load argument specs for role '%s'" % role_name)
+
+        loader = DataLoader()
+        return loader.load_from_file(path, cache=False, unsafe=True)
+
+    def _find_all_normal_roles(self, role_paths, name_filters=None):
+        """Find all non-collection roles that have an argument spec file.
+
+        :param role_paths: A tuple of one or more role paths.
+        :param name_filters: A tuple of one or more role names used to filter the results.
+
+        :returns: A set of tuples consisting of: role name, full role path
+        """
+        found = set()
+        for path in role_paths:
+            if not os.path.isdir(path):
+                display.warning("Invalid role path: {0}".format(path))
+            else:
+                # Check each subdir for a meta/argument_specs.yml file
+                for entry in os.listdir(path):
+                    role_path = os.path.join(path, entry)
+                    full_path = os.path.join(role_path, 'meta', self.ROLE_ARGSPEC_FILE)
+                    if os.path.exists(full_path):
+                        if name_filters is None or entry in name_filters:
+                            found.add((entry, role_path))
+        return found
+
+    def _find_all_collection_roles(self, name_filters=None, collection_filter=None):
+        """Find all collection roles with an argument spec.
+
+        :param name_filters: A tuple of one or more role names used to filter the results. These
+            might be fully qualified with the collection name (e.g., community.general.roleA)
+            or not (e.g., roleA).
+
+        :param collection_filter: A string containing the FQCN of a collection which will be
+            used to limit results. This filter will take precedence over the name_filters.
+
+        :returns: A set of tuples consisting of: role name, collection name, collection path
+        """
+        found = set()
+        b_colldirs = list_collection_dirs()
```

Comment:
Passing the collection filter here rather than checking below will prevent importing unrelated collections in `_get_collection_name_from_path`.

Reply:
Resolved in latest commits.